### ISSUE TITLE
Reverse Marshal with powershell

### ISSUE THREAD

**Issue by cavey1986** [bongiovimatthew-microsoft/pscredentialWithCert#2]
Is there a way you could assist with the code in powershell format to reverse marshal the username field that is generated?  I have been trying to take the C# code and work it into powershell but can't seem to get it to work.  I am creating a Burner application that requires Two Person Integrity and the string that is pulled from the Username field during PS-Credential does not help identify who actually authenticated.  Is just shows the @@*********************************, thanks for any assistance you can provide.

**Comment by bongiovimatthew-microsoft** [comment]
@cavey1986, if you send the code attempt you have so far, I can take a look and try to help out

**Comment by bongiovimatthew-microsoft** [comment]
Actually, it looks like most of the logic can be left as C# code, and wrapped using the PowerShell `Add-Type` cmdlet. I am able to successfully use the [code I add here](https://github.com/bongiovimatthew-microsoft/pscredentialWithCert/blob/powershellPort/PowerShell/ReverseMarshal.ps1). Let me know if that works.

**Comment by cavey1986** [comment]
Matt,

Thanks for responding so quickly, I will add this code into what I have and determine if I can get it to work.  If I have issues, I will provide the code I have created and see if you can assist further.  I appreciate your assistance sir.

**Comment by cavey1986** [comment]
Matt,

The code that you gave me works perfectly so thanks again for providing that.  I am now trying to figure out and if/else process to determine if the username field is populated with a standard username format, then provide that feedback to the textbox within the app.  If a certificate is provided for the username, reverse marshal with the code you provided and list the SubjectName.Name as the user within the text in the app.  Im trying to not have to create two different applications and have the logic to determine what has been entered as the username and process the information accordingly.  Below is an example of what I have but can't get it to work.  Im sure it has something to do with the string that is provided by the marshalling for the certificate information.  

$script:cred = Get-Credential -Message "Enter User 1 credentials"
$script:user1 = $cred.UserName

if ($user1 -notcontains "@@")
{
	Write-Host "Username is $user1"
}
else
{
	DecryptCertInfo
	Write-Host "Username is $certname"
}

**Comment by bongiovimatthew-microsoft** [comment]
@cavey1986, I updated [the code](https://github.com/bongiovimatthew-microsoft/pscredentialWithCert/blob/powershellPort/PowerShell/ReverseMarshal.ps1) with a function that I use to check if the username field is populated with a certificate or username. The `CredUnmarshalCredential` function actually supports this. In the code I gave you, the check is broken into two separate functions, which means that you are calling `CredUnmarshalCredential` twice. For your project, you will probably want to re-work the code a bit to combine these two functions so that you don't have to double-call `CredUnmarshalCredential`.
